### PR TITLE
Sync metadata implementation with DataAPI.jl 1.12.0

### PR DESCRIPTION
**Pull request by bkamins**
This PR should be passed through CI after https://github.com/JuliaData/DataAPI.jl/pull/53 is merged and DataAPI.jl 1.12.0 is released. Other than that it should be complete (not tested, but the changes are minor - mostly simplification of the code because of dropped default implementations + adding support for `default`).

### CODE REVIEW

File: src/other/metadata.jl [hunk 1]
```diff
@@ -82,22 +99,29 @@ See also: [`metadatakeys`](@ref), [`metadata!`](@ref),
...
+        end
     end
     return style ? meta[key] : meta[key][1]
 end
 
-metadata(x::Union{DataFrameRows, DataFrameColumns},
-         key::AbstractString; style::Bool=false) =
-    metadata(parent(x), key, style=style)
+metadata(x::Union{DataFrameRows, DataFrameColumns}, key::AbstractString,
+                  default=MetadataMissingDefault(); style::Bool=false) =
```
Comment (nalimilan) [inline src/other/metadata.jl:116]:
```suggestion
metadata(x::Union{DataFrameRows, DataFrameColumns}, key::AbstractString,
         default=MetadataMissingDefault(); style::Bool=false) =
```

File: src/other/metadata.jl [hunk 2]
```diff
@@ -311,14 +335,19 @@ function emptymetadata!(x::Union{DataFrameRow, SubDataFrame})
...
 
 Return column-level metadata value associated with `df` for column `col` and key `key`.
 
-`SubDataFrame` and `DataFrameRow` expose only `:note`-style metadata of their parent.
+`SubDataFrame` and `DataFrameRow` expose only `:note`-style metadata of their parent
+ (except if `default` is passed and `key` for column `col` does not exist).
+
+If `default` is passed then return it if reading metadata is supported but
+mapping for `key` in column `col` is missing. If `style=true` return `(default, :default)`.
+If `col` does not exist in `df` always throw an error.
```
Comment (nalimilan) [inline src/other/metadata.jl:350]:
```suggestion
If `default` is passed then return it if `key` does not exist for column `col`;
if `style=true` return `(default, :default)`.
If `col` does not exist in `df` always throw an error.
```

File: src/other/metadata.jl [hunk 3]
```diff
@@ -311,14 +335,19 @@ function emptymetadata!(x::Union{DataFrameRow, SubDataFrame})
...
+    colmetadata(df::AbstractDataFrame, col::ColumnIndex, key::AbstractString, [default]; style::Bool=false)
+    colmetadata(dfr::DataFrameRow, col::ColumnIndex, key::AbstractString, [default]; style::Bool=false)
+    colmetadata(dfc::DataFrameColumns, col::ColumnIndex, key::AbstractString, [default]; style::Bool=false)
+    colmetadata(dfr::DataFrameRows, col::ColumnIndex, key::AbstractString, [default]; style::Bool=false)
 
 Return column-level metadata value associated with `df` for column `col` and key `key`.
 
-`SubDataFrame` and `DataFrameRow` expose only `:note`-style metadata of their parent.
+`SubDataFrame` and `DataFrameRow` expose only `:note`-style metadata of their parent
+ (except if `default` is passed and `key` for column `col` does not exist).
```
Comment (nalimilan) [inline src/other/metadata.jl:346]:
```suggestion
`SubDataFrame` and `DataFrameRow` expose only `:note`-style metadata of their parent.
```

File: src/other/metadata.jl [hunk 4]
```diff
@@ -63,16 +77,19 @@ const COLMETADATA_EXAMPLE =
...
+    metadata(dfr::DataFrameRow, key::AbstractString, [default]; style::Bool=false)
+    metadata(dfc::DataFrameColumns, key::AbstractString, [default]; style::Bool=false)
+    metadata(dfr::DataFrameRows, key::AbstractString, [default]; style::Bool=false)
 
 Return table-level metadata value associated with `df` for key `key`.
 If `style=true` return a tuple of metadata value and metadata style.
 
 `SubDataFrame` and `DataFrameRow` expose only `:note`-style metadata of their
-parent.
+parent (except if `default` is passed and `key` does not exist).
```
Comment (nalimilan) [inline src/other/metadata.jl:89]: I don't think this precision is needed, as non-`:note`-style metadata is not exposed when `default` is provided either.
```suggestion
parent.
```

File: src/other/metadata.jl [hunk 5]
```diff
@@ -63,16 +77,19 @@ const COLMETADATA_EXAMPLE =
...
+
+If `default` is passed then return it if reading metadata is supported but
+mapping for `key` is missing. If `style=true` return `(default, :default)`.
```
Comment (nalimilan) [inline src/other/metadata.jl:92]:
```suggestion
If `default` is passed then return it if `key` does not exist;
if `style=true` return `(default, :default)`.
```

File: src/other/metadata.jl [hunk 6]
```diff
@@ -328,49 +357,48 @@ See also: [`metadata`](@ref), [`metadatakeys`](@ref),
...
 end
 
-# here and similar definitions below are added to avoid dispatch ambiguity
-for T in (Symbol, ColumnIndex)
-    @eval colmetadata(df::DataFrame, col::$T, key::AbstractString; style::Bool=false) =
-        colmetadata(df, Int(index(df)[col]), key, style=style)
-end
+colmetadata(x::Union{DataFrameRows, DataFrameColumns}, col::ColumnIndex,
+            key::AbstractString, default=MetadataMissingDefault(); style::Bool=false) =
+        colmetadata(parent(x), col, key, default; style=style)
```
Comment (nalimilan) [inline src/other/metadata.jl:387]:
```suggestion
    colmetadata(parent(x), col, key, default; style=style)
```

File: src/other/metadata.jl [hunk 7]
```diff
@@ -328,49 +357,48 @@ See also: [`metadata`](@ref), [`metadatakeys`](@ref),
...
-        colname = names(df)[idx]
-        throw(ArgumentError("\"$key\" not found in column-level metadata for column \"$colname\""))
+        if default === MetadataMissingDefault()
+            colname = names(df)[idx]
+            throw(ArgumentError("\"$key\" not found in column-level metadata for column \"$colname\""))
+        else
+            return style ? (default, :default) : default
+        end
     end
     return style ? col_meta[key] : col_meta[key][1]
```
Comment (nalimilan) [inline src/other/metadata.jl:382]:
```suggestion
```

File: src/other/metadata.jl [hunk 8]
```diff
@@ -396,8 +424,8 @@ See also: [`metadata`](@ref), [`metadatakeys`](@ref),
 $COLMETADATA_EXAMPLE
 ```
 """
-function colmetadatakeys(df::DataFrame, col::Int)
-    idx = index(df)[col] # bounds checking
+function colmetadatakeys(df::DataFrame, col::ColumnIndex)
+    idx = index(df)[col] # check if column exists and get its integer number
```
Comment (nalimilan) [inline src/other/metadata.jl:428]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl [hunk 9]
```diff
@@ -472,8 +486,8 @@ See also: [`metadata`](@ref), [`metadatakeys`](@ref),
 $COLMETADATA_EXAMPLE
 ```
 """
-function colmetadata!(df::DataFrame, col::Int, key::AbstractString, value::Any; style)
-    idx = index(df)[col] # bounds checking
+function colmetadata!(df::DataFrame, col::ColumnIndex, key::AbstractString, value::Any; style)
+    idx = index(df)[col] # check if column exists and get its integer number
```
Comment (nalimilan) [inline src/other/metadata.jl:490]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl [hunk 10]
```diff
@@ -489,44 +503,30 @@ function colmetadata!(df::DataFrame, col::Int, key::AbstractString, value::Any;
...
+    return x
 end
 
-for T1 in (DataFrameRows, DataFrameColumns), T2 in (Int, Symbol, ColumnIndex)
-    @eval function colmetadata!(x::$T1,
-                                col::$T2, key::AbstractString, value::Any; style)
-        colmetadata!(parent(x), col, key, value; style=style)
-        return x
+function colmetadata!(x::Union{DataFrameRow, SubDataFrame},
+                    col::ColumnIndex, key::AbstractString, value::Any; style)
```
Comment (nalimilan) [inline src/other/metadata.jl:513]:
```suggestion
function colmetadata!(x::Union{DataFrameRow, SubDataFrame},
                      col::ColumnIndex, key::AbstractString, value::Any; style)
```

File: src/other/metadata.jl [hunk 11]
```diff
@@ -489,44 +503,30 @@ function colmetadata!(df::DataFrame, col::Int, key::AbstractString, value::Any;
...
-                                "that is already present in the parent and does not " *
-                                "have :note style is not allowed"))
-        end
-        colmetadata!(df, idx, key, value, style=style)
-        return x
+    df = parent(x)
+    cols_meta = getfield(df, :colmetadata)
+    idx = index(df)[col_name]
+    if cols_meta !== nothing && haskey(cols_meta, idx) &&
+    haskey(cols_meta[idx], key) && cols_meta[idx][key][2] !== :note
```
Comment (nalimilan) [inline src/other/metadata.jl:523]:
```suggestion
        haskey(cols_meta[idx], key) && cols_meta[idx][key][2] !== :note
```

File: src/other/metadata.jl [hunk 12]
```diff
@@ -547,8 +547,8 @@ See also: [`metadata`](@ref), [`metadatakeys`](@ref),
 
 $COLMETADATA_EXAMPLE
 """
-function deletecolmetadata!(df::DataFrame, col::Int, key::AbstractString)
-    idx = index(df)[col] # bounds checking
+function deletecolmetadata!(df::DataFrame, col::ColumnIndex, key::AbstractString)
+    idx = index(df)[col] # check if column exists and get its integer number
```
Comment (nalimilan) [inline src/other/metadata.jl:551]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl [hunk 13]
```diff
@@ -636,54 +624,40 @@ julia> colmetadatakeys(df)
 ()
 ```
 """
-function emptycolmetadata!(df::DataFrame, col::Int)
-    idx = index(df)[col] # bounds checking
+function emptycolmetadata!(df::DataFrame, col::ColumnIndex)
+    idx = index(df)[col] # check if column exists and get its integer number
```
Comment (nalimilan) [inline src/other/metadata.jl:628]:
```suggestion
    idx = index(df)[col] # check if column exists and get its integer index
```

File: src/other/metadata.jl [hunk 14]
```diff
@@ -745,11 +722,14 @@ end
...
-        val, style = metadata(src, key, style=true)
-        metadata!(dst, key, val, style=style)
+    if DataAPI.metadatasupport(typeof(src)).read
+        for key in metadatakeys(src)
+            val, style = metadata(src, key, style=true)
+            metadata!(dst, key, val, style=style)
+        end
     end
     emptycolmetadata!(dst)
+    DataAPI.colmetadatasupport(typeof(src)).read || return nothing
```
Comment (nalimilan) [inline src/other/metadata.jl:732]: Use an `if` for consistency with above?

File: test/metadata.jl [hunk 15]
```diff
@@ -24,15 +24,69 @@ function check_allnotemetadata(x::Union{AbstractDataFrame,
     return true
 end
 
+@testset "metadata support" begin
```
Comment (nalimilan) [inline test/metadata.jl:27]:
```suggestion
@testset "metadatasupport" begin
```

**Review by nalimilan — APPROVED** [review]

**Comment by nalimilan** [comment]
Nice cleanup!

File: src/other/metadata.jl (hunk 14 above)
Comment (bkamins) [reply src/other/metadata.jl]: OK. I have changed short-circut to `if`s everywhere then. After CI passes I will merge this PR.
Then I will wait one day to wait for final responses, and will make a release of DataFrames.jl 1.4 tomorrow if there are none.

File: src/other/metadata.jl [hunk 16]
```diff
@@ -1,5 +1,19 @@
 ### Metadata API from DataAPI.jl
 
+# private type that is passed as a default value in metadata and colmetadata
+# do detect the fact that no default was passed
```
Comment (bkamins) [inline src/other/metadata.jl:4]:
```suggestion
# to detect the fact that no default was passed
```

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**